### DISCUSSION TITLE
what the difference between StateNotifierProvider and NotifierProvider?

### DISCUSSION THREAD

**Discussion by bigdogs** [rrousselGit/riverpod#2380] [General]
hi, 
what the difference between StateNotifierProvider and NotifierProvider?  I found that their document and example demo seem extacly same.... 
https://docs-v2.riverpod.dev/docs/providers/notifier_provider 
https://docs-v2.riverpod.dev/docs/providers/state_notifier_provider

Docuemnt said `NotifierProvider` is preferred.. But  the constructor fuction of  NotifierProvider doesn't have `WidgetRef` which means I can't observer other provider use `NotifierProvider`?

I'm really consufused :(

**Comment by ArunPrakashG** [comment]
Ref is available in Notifier & AsyncNotifier providers by default. You don't have to pass it via the constructor.

**Reply by bigdogs** [reply]
> Ref is available in Notifier & AsyncNotifier providers by default. You don't have to pass it via the constructor.

sorry for my poor English. I mean argument of  the functor to create the `NotifierProvider` is not a `WidgetRef`, then I can't `watch` other provider ...  
![image](https://user-images.githubusercontent.com/20614926/227204624-3a586273-ca95-43a5-a7bb-e7145716b432.png)

`StateNotifierProvider` can watch other provider, but the document said it's not preferred, and didn't say why..  
![image](https://user-images.githubusercontent.com/20614926/227205843-386179aa-00cf-44b2-a6cd-3143dd74b3c9.png)

I have no idea which one should be used

**Reply by Shreemanarjun** [reply]
You can use ref.watch(anyProvider) inside **Notifier** class (**ExplorerProvider**) .
